### ISSUE TITLE
Find neighbors based on edge attributes

### ISSUE THREAD

**Issue by sushantdhiman** [graphology/graphology#351]
**Background**

Using this library I often need to this "find all neighbors, connected with special kind of edge (defined as kind field on edge attribute)". Currently I am using `reduceEdges`, something like this

```js
   graph.reduceEdges(
      source,
      (acc, _, attributes, source, target) => {
        if (attributes.kind === 'special_kind') {
          acc.push(target);
          acc.push(source);
        }
        return acc;
      },
      [] as GraphNode[]
    )
```

I would also like to point out that I am using mix of directed and undirected edges.

**Feature Request**

Callback function from neighbor traversal [methods](https://graphology.github.io/iteration.html#neighbors) should include edge attributes as third argument (where applicable). This will allow skipping / including neighbors based on their edge attributes.

**Comment by Yomguithereal** [comment]
Hello @sushantdhiman,

> Callback function from neighbor traversal methods should include edge attributes as third argument (where applicable). This will allow skipping / including neighbors based on their edge attributes.

This is not possible because the method much work with all kind of graphs (mixed and multi, notably), and the invariant this method must respect is to ensure you will only see each neighbor only once. 

What's wrong with using edge methods (`reduceEdges` or `filterEdges` as you do) for this kind of thing?

**Comment by sushantdhiman** [comment]
> What's wrong with using edge methods (reduceEdges or filterEdges as you do) for this kind of thing (you can use a Set as reduce or foreach target to deduplicate neighbors while iterating)?

There is nothing wrong with this, I was just exploring possibility of some official helper for this.

You are right that for neighbor traversal each neighbor must appear only once, given that graph could be a multi graph. A possible solution could be to add `edges[]` argument to neighbor callback function. This will expose information about edges linked to that neighbor. I think this will be valuable addition, for many cases it is important to know how neighbor is connected.

If you think this addition is not worth it, you may close this issue.

**Comment by Yomguithereal** [comment]
@sushantdhiman this would be a valuable addition to have an helper for this kind of scenario indeed but I would not want it from the data structure itself because it has a performance cost the general case should not have to pay. What's more, it can be implemented in userland without any issue. We could add this to `graphology-utils` for instance. Something along the line of `forEachGroupedEdges` or `forEachEdgeGroup`. What do you think?

**Comment by micahscopes** [comment]
This would be really useful.  Right now using `reduceEdges` to do this adds a good amount of mental overhead.

**Comment by Yomguithereal** [comment]
To be sure, we are speaking of something which is only useful when dealing with mixed and/or multigraphs? I see @sushantdhiman use case but can you develop yours @micahscopes, as it could be useful to better design the utility function if required?

**Comment by micahscopes** [comment]
@Yomguithereal sure, my usecase is basically the same. I'd like to do various walks on (multi)graphs that are informed both by edge properties and node properties. `reduceEdges` is basically fine, but something a little more ergonomic would be awesome.

Another alternative would be having the option to somehow select a subgraph (yielding a new Graph object maybe?) and then use the neighbors functions on that prefiltered selection.

I'm new to this library so I don't know how hard that would be or if it would be consistent with the design.

**Comment by Yomguithereal** [comment]
> Another alternative would be having the option to somehow select a subgraph (yielding a new Graph object maybe?) and then use the neighbors functions on that prefiltered selection.

There is a `subgraph` function in [graphology-operators](https://graphology.github.io/standard-library/operators#subgraph) package. But this would not be an efficient way to deal with this kind of scenarios.

I want to add something to `graphology-utils` in the lines of `#.forEachNeighborWithRelatedEdges` or something of the kind but I don't want it as a argument of the neighbor methods because it is a very costly thing to do by default and that most users don't use. What's more, there are different optimization strategies to be used if you want to iterate over the whole graph's adjacency or just do it on specific nodes.

**Comment by Yomguithereal** [comment]
Another issue I have here is about the grouping of the edges. Some people might want an array/set of edge keys, some might want an array of full edge entries directly, some might want to filter out some edges or aggregate them. This is not an easy thing to do design-wise and I am trying to think a little bit more about this.

**Comment by micahscopes** [comment]
Yeah I guess the simplest could be something like `filterEdges` but which returns a list of neighbor keys rather than edge keys, and with a simplified callback function. Maybe something with an API like this: 

```
filterEdgeNeighbors(
  startNodeKey,
  (edgeAttributes, neighborAttributes, direction) => { ... }
)
```
